Issue #9950: Ordered Aggregate Performance

hawkfish opened the pull request:
Reduce memory requirements and increase performance.

* Add linked list merging
* Add Swap to reduce vector overhead
* Fix PhysicalUngroupedAggregate to manage ArenaAllocator lifetimes.

fixes: #9980
fixes: duckdblabs/duckdb-internal#892

Commit:
Issue #5920: Ordered Aggregate Sorting

Attempt to fix bogus reuse after move tidy failure.

Commit:
Merge branch 'master' into sorted-agg

Commit:
Merge branch 'master' into sorted-agg

Commit:
Issue #5920: Ordered Aggregate Sorting

Cleanup and better chunk reuse.

Commit:
Merge branch 'main' into sorted-agg

Commit:
Issue #9950: Array List Segments

The list segment code for arrays was incorrectly leaving out blank rows for NULL values.

Commit:
Issue #9950: Ordered Aggregate Performance

Checkpoint commit that is almost working.

hawkfish committed:
Merge branch 'main' into sorted-agg

hawkfish committed:
Issue #9950: Ordered Aggregate Performance

Fix memory management issues.

hawkfish committed:
Issue #9950: Ordered Aggregate Performance

Release memory after the contents have been absorbed.

hawkfish committed:
Merge branch 'main' into sorted-agg

hawkfish committed:
Issue #9950: Ordered Aggregate Performance

* Add linked list merging
* Fix PhysicalUngroupedAggregate to share one ArenaAllocator between Sink threads.

hawkfish committed:
Issue #9950: Ordered Aggregate Performance

* Add Swap to reduce vector overhead
* Fix PhysicalUngroupedAggregate to manage ArenaAllocator lifetimes.

hawkfish committed:
Merge branch 'main' into sorted-agg

lnkuiper reviewed (approved):
LGTM. Great that you were able to solve the problem!

I'd like to look into more optimizations in the future (as discussed in the internal issue), when I have the time :)

Mytherin reviewed (commented):
Thanks! LGTM - one minor comment:

hawkfish commented on benchmark/micro/aggregate/ordered_first.benchmark:
```diff
@@ -16,4 +16,4 @@ SELECT SUM(agg) FROM (
 
 result I
-24420932461
+49999995000000
```
Wow - that looks wrong. And I have no memory of changing it!

hawkfish replied:
Oh the test is just wrong.

Commit:
Issue #9950: Correct Benchmark Result

Also make it multithreaded.
The result been checked against an equivalent MAX query.

Mytherin commented:
Thanks!
